packages/@lwc/template-compiler/src/shared/ir.ts

```diff
@@ -42,105 +141,293 @@ export function createElement(original: parse5.Element): IRElement {
         throw new Error('Invalid element AST node. Missing source code location.');
     }
 
+    return location;
+}
+
+export function parseSourceLocation(location: parse5.Location): SourceLocation {
     return {
-        type: 'element',
-        tag: original.tagName,
-        namespace: original.namespaceURI,
+        startLine: location.startLine,
+        startColumn: location.startCol,
+        endLine: location.endLine,
+        endColumn: location.endCol,
+        start: location.startOffset,
+        end: location.endOffset,
+    };
+}
+
+export function literal(value: string | boolean): Literal {
+    return {
+        type: LWCNodeType.Literal,
+        value,
+    };
+}
+
+export function forEach(
+    name: string,
+    expression: Expression,
+    location: SourceLocation,
+    item: Identifier,
+    index?: Identifier
+): ForEach {
+    return {
+        type: LWCNodeType.ForEach,
+        name,
```

Comment:
`ForEach`, `If` and `ForOf` AST nodes are pure control flow nodes. Those shouldn't have a `name` property. 
Looking at the interfaces, the `BaseParentNode` shouldn't have a name property.